core/embed/rust/src/micropython/buffer.rs

```diff
@@ -27,22 +32,48 @@ impl StrBuffer {
     }
 
     pub fn alloc(val: &str) -> Result<Self, Error> {
+        unsafe {
+            // SAFETY: Copying existing UTF-8 buffer of the right length.
+            Self::alloc_with(val.len(), |buffer| {
+                // SAFETY: Memory should be freshly allocated and as such cannot overlap.
+                ptr::copy_nonoverlapping(val.as_ptr(), buffer.as_mut_ptr(), buffer.len());
+                // SAFETY: Copy of valid UTF-8 string.
+                str::from_utf8_unchecked(slice::from_raw_parts(buffer.as_ptr(), buffer.len()))
+            })
+        }
+    }
+
+    /// SAFETY: The caller must ensure that `func` returns a slice that:
+    /// - spans the entire length of the given buffer, and
+    /// - contains valid UTF-8.
```

Comment:
This comment implies that `func` should be of type `-> ()`.
The return value is useless anyway because we only assert that it covers the whole slice and immediately throw it away. 

What we _could_ do is have it return `usize` so that `ret <= len`, allowing allocation above needed size. And if we wanted to make this function fully safe, we could then call `str::from_utf8`. ISTM we can afford that because it only triggers relatively rarely. Or we can have an unsafe internal helper used by both `alloc` and `alloc_with` that skips the validation, if this becomes a problem.